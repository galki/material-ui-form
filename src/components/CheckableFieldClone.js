// @flow

import React from 'react'

import Checkbox from 'material-ui/Checkbox'
import Switch from 'material-ui/Switch'


type Props = {
  field?: Object,
  fieldComp: Object,
  onConstruct: Function,
  onToggle: Function,
};

type State = {
  checked: boolean,
};

export default class CheckableFieldClone extends React.Component<Props, State> {
  static defaultProps = {
    field: {},
  }

  constructor(props: Object) {
    super(props)
    const { fieldComp } = props

    if (![Checkbox, Switch].includes(fieldComp.type)) {
      throw new Error('CheckableFieldClone should be a Checkbox or Switch')
    }
    if (fieldComp.props.name === undefined || fieldComp.props.value === undefined) {
      throw new Error('CheckableFieldClone name and value must be defined')
    }

    let checked = props.field.value
    if (props.field.value === undefined) {
      checked = fieldComp.props.checked || false
      this.props.onConstruct(fieldComp.props)
    }
    this.state = { checked }
  }

  onToggle = (event: Event, checked: boolean) => {
    const { fieldComp, fieldComp: { props: { name, value } } } = this.props
    this.setState({ checked })
    this.props.onToggle(name, value, checked)
    if (fieldComp.props.onChange !== undefined) {
      fieldComp.props.onChange(checked, { name, value }, event)
    }
  }

  render() {
    const { fieldComp } = this.props
    return React.cloneElement(fieldComp, {
      value: fieldComp.props.value,
      checked: this.state.checked,
      onChange: this.onToggle,
    })
  }
}
