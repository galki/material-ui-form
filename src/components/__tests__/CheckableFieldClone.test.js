import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Checkbox from 'material-ui/Checkbox'

import CheckableFieldClone from '../CheckableFieldClone'


const field = {
  isPristine: true,
  isRequired: null,
  pristineValue: null,
  validations: [],
  validators: [],
  value: undefined,
}

describe('<CheckableFieldClone>:<Checkbox>', () => {
  const wrapper = shallow(
    <CheckableFieldClone
      field={field}
      fieldComp={(<Checkbox
        onChange={jest.fn()}
        checked
        name="foo"
        value="bar"
      />)}
      onConstruct={jest.fn()}
      onToggle={jest.fn()}
    />
  )

  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should handle onToggle events', () => {
    const checked = false
    const event = { target: { checked } }
    wrapper.simulate('change', event)
    expect(wrapper.instance().props.onToggle).toHaveBeenCalled()
  })
})

describe('<CheckableFieldClone> Invalid props', () => {
  it('should throw if control type is other than Checkbox or Switch', () => {
    let error
    try {
      shallow(
        <CheckableFieldClone
          field={field}
          fieldComp={<input />}
          onConstruct={jest.fn()}
          onToggle={jest.fn()}
        />
      )
    } catch (e) {
      error = e
    }
    expect(error).toBeInstanceOf(Error)
  })

  it('should throw if control type fieldComp has no name or value props', () => {
    let error
    try {
      shallow(
        <CheckableFieldClone
          field={field}
          fieldComp={(<Checkbox
            onChange={jest.fn()}
            checked
          />)}
          onConstruct={jest.fn()}
          onToggle={jest.fn()}
        />
      )
    } catch (e) {
      error = e
    }
    expect(error).toBeInstanceOf(Error)
  })
})
