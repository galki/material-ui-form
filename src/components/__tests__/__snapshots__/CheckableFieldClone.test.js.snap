// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckableFieldClone>:<Checkbox> should render 1`] = `
<WithStyles(Checkbox)
  checked={true}
  name="foo"
  onChange={[Function]}
  value="bar"
/>
`;
